minifi_main/AgentDocs.cpp

```diff
@@ -64,130 +65,145 @@ std::string formatAllowedValues(const minifi::core::Property& property) {
-        dynamic_properties.addRow({
-            formatName(dynamic_property.name, false),
-            std::string(dynamic_property.value),
-            formatDescription(dynamic_property.description, dynamic_property.supports_expression_language)
-        });
-      }
-      outfile << dynamic_properties.toString() << '\n';
-    }
+  docs << dynamic_properties.toString();
+}
 
-    outfile << "### Relationships\n\n";
-    Table relationships{{"Name", "Description"}};
-    for (const auto &rel : processor.second.class_relationships_) {
-      relationships.addRow({rel.getName(), formatDescription(rel.getDescription())});
-    }
-    outfile << relationships.toString() << '\n';
-
-    if (!processor.second.output_attributes_.empty()) {
-      outfile << "### Output Attributes\n\n";
-      Table output_attributes{{"Attribute", "Relationship", "Description"}};
-      for (const auto& output_attribute : processor.second.output_attributes_) {
-        output_attributes.addRow({
-            std::string(output_attribute.name),
-            formatListOfRelationships(output_attribute.relationships),
-            formatDescription(output_attribute.description)});
-      }
-      outfile << output_attributes.toString() << '\n';
-    }
+void writeRelationships(std::ostream& docs, const minifi::ClassDescription& documentation) {
+  docs << "\n### Relationships\n\n";
+  minifi::docs::Table relationships{{"Name", "Description"}};
+  for (const auto &rel : documentation.class_relationships_) {
+    relationships.addRow({rel.getName(), formatDescription(rel.getDescription())});
   }
+  docs << relationships.toString();
+}
+
+void writeOutputAttributes(std::ostream& docs, const minifi::ClassDescription& documentation) {
+  if (documentation.output_attributes_.empty()) { return; }
+
+  docs << "\n### Output Attributes";
+  minifi::docs::Table output_attributes{{"Attribute", "Relationship", "Description"}};
+  for (const auto &output_attribute : documentation.output_attributes_) {
+    output_attributes.addRow({
+        std::string(output_attribute.name),
+        formatListOfRelationships(output_attribute.relationships),
+        formatDescription(output_attribute.description)});
+  }
+  docs << "\n\n" << output_attributes.toString();
+}
 
-  std::map<std::string, std::filesystem::path> fileList;
-  auto fileFind = [&fileList](const std::filesystem::path& base_path, const std::filesystem::path& file) -> bool {
-    if (file.string().find(".extra") == std::string::npos) {
-      auto file_name = file.string();
-      ranges::actions::transform(file_name, [](auto ch) { return ::tolower(static_cast<unsigned char>(ch)); });
-      fileList.emplace(file_name, base_path / file);
+std::string extractClassName(const std::string& full_class_name) {
+  return minifi::utils::StringUtils::split(full_class_name, ".").back();
+}
+
+constexpr auto LowercaseFirst = [](const auto& key_value) {
+  return minifi::utils::StringUtils::toLower(key_value.first);
+};
```

Comment:
I'd apply this function to its own name. :)
Just to be consistent with the naming of other functions. I also wonder why is it a function object: a simple function can also be `constexpr`, and be passed by its address to `sort`.
```suggestion
constexpr auto lowercaseFirst = [](const auto& key_value) {
  return minifi::utils::StringUtils::toLower(key_value.first);
};
```

Reply:
It's a constant (of type lambda), not a function, and I think our convention is to use UpperCamelCase for these.

The reason it's not a function is that in that in that case I'd have to specify the argument type (which is quite long), as it won't get deduced automatically.  So I could do
```c++
std::string lowercaseFirst(const std::pair<std::string, minifi::ClassDescription>& key_value) {
  return minifi::utils::StringUtils::toLower(key_value.first);
}
```
Is that better?

Reply:
I'd still consider it a function, despite being implemented as an empty constant function object. But if you disagree, that's fine.

You can use `auto` in non-lambda functions as well to deduce the argument type, since C++20 (and previously as a GCC extension). I don't really care about the syntactic details of a function definition, whether it's a function object or a regular function. I'd prefer a lowercase starting letter, and it looks like a regular function with that could be a compromise between our viewpoints: it's no longer a function object, so the function naming scheme applies without a doubt, and you can still use `auto` parameter type, unless some compiler has issues with it.
```suggestion
std::string lowercaseFirst(const auto& key_value) {
  return minifi::utils::StringUtils::toLower(key_value.first);
};
```

Reply:
The problem is with the sort: with your latest version,
```c++
  ranges::sort(controller_services, std::less(), lowercaseFirst);
```
doesn't work.  I don't know why it can deduce the template parameter for `less` but not for `lowercaseFirst`, but it can't.  You can experiment with it here: https://godbolt.org/z/Gx4vs5K47

This works:
```c++
  ranges::sort(controller_services, std::less(), lowercaseFirst<std::pair<std::string, minifi::ClassDescription>>);
```
but then I'd might as well use `<std::pair<std::string, minifi::ClassDescription>>` instead of `auto` in the function definition.  I have done that in 73c477c054b5336d63cb916a739d1875d61763b1; let me know if you want to change it to something else.

Reply:
Yeah, you're right, taking the address of a generic function doesn't really work. `std::less` works, because it's a function object, and the template argument deduction is done at the call operator. But with function pointers, you can't have a pointer to a function that hasn't been instantiated yet, it has to be a template specialization. My compiler explorer link: https://godbolt.org/z/bsP8KjG7T

I would've written the same as your initial version, except for the identifier casing. I'm OK with either version, we can leave it as is.